packages/chatbot-server-mongodb-public/src/tracing.ts

```diff
@@ -76,6 +94,128 @@ export function extractTracingData(messages: SomeMessage[]) {
     isVerifiedAnswer,
     llmDoesNotKnow,
     numRetrievedChunks,
+    latestUserMessage,
+    latestAssistantMessage,
+  };
+}
+
+function getTracingScores(
+  tracingData: ReturnType<typeof extractTracingData>,
+  k: number
+) {
+  return {
+    RejectedQuery: tracingData.rejectQuery === true ? 1 : null,
+    VerifiedAnswer: tracingData.isVerifiedAnswer === true ? 1 : null,
+    LlmDoesNotKnow: tracingData.llmDoesNotKnow === true ? 1 : null,
+    [`RetrievedChunksOver${k}`]:
+      tracingData.isVerifiedAnswer !== true
+        ? tracingData.numRetrievedChunks / k
+        : null,
+  };
+}
+
+interface LlmAsAJudge {
+  /**
+    Percent of numbers to judge. Between 0 and 1.
+   */
+  percentToJudge: number;
+  judgeModel: string;
+  judgeEmbeddingModel: string;
+  openAiConfig: Pick<
+    Parameters<typeof Faithfulness>[0],
+    | "azureOpenAi"
+    | "openAiApiKey"
+    | "openAiBaseUrl"
+    | "openAiDangerouslyAllowBrowser"
+    | "openAiDefaultHeaders"
+    | "openAiOrganizationId"
+  >;
+}
+const nullScore = {
+  score: null,
+};
+
+export async function getLlmAsAJudgeScores(
+  {
+    judgeModel,
+    openAiConfig,
+    judgeEmbeddingModel,
+  }: Omit<LlmAsAJudge, "percentToJudge">,
+  tracingData: ReturnType<typeof extractTracingData>
+) {
+  // Return if we don't have the necessary data to judge
+  if (
+    tracingData.isVerifiedAnswer ||
+    tracingData.rejectQuery ||
+    tracingData.numRetrievedChunks === 0 ||
+    tracingData.llmDoesNotKnow ||
+    !tracingData.latestUserMessage ||
+    !tracingData.latestAssistantMessage
+  ) {
+    return;
+  }
+  const input = tracingData.latestUserMessage.content;
+  const output = tracingData.latestAssistantMessage.content;
+  const context = tracingData.latestUserMessage.contextContent
+    ?.map((c) => c.text)
+    // Have to do this type casting to make it compile. Unclear why.
+    .filter((c) => typeof c === "string") as string[] | undefined;
```

Comment:
```suggestion
    .filter((c): c is string => typeof c === "string");
```

I think using a type predicate (i.e. the `is` keyword) should fix it without a cast

Reply:
ok, i def still dont grok `is`, but will see if this gets it

Reply:
tl;dr - `is` lets you return a `boolean` that also conveys type information.

This is necessary to get types out of `Array.prototype.filter` because JS is a crazy language.